Ensure uniqueness of parsed SequenceContainer objects

medley56 opened the pull request:
Modify the parsing of XTCE documents to ensure that SequenceContainer objects are kept unique and reused instead of parsed and stored multiple times.

This PR adds checks to the parsing of ParameterSet, ParemeterTypeSet, and ContainerSet to ensure that the objects we parse out of those elements are uniquely represented in the resulting definition object. We also issue warnings for parameters and parameter types that are defined but not used.

## Checklist
- [x] Changes are fully implemented without dangling issues or TODO items
- [x] Deprecated/superseded code is removed or marked with deprecation warning
- [x] Current dependencies have been properly specified and old dependencies removed
- [x] New code/functionality has accompanying tests and any old tests have been updated to match any new assumptions
- [x] The changelog.md has been updated

greglucas commented on space_packet_parser/xtce/containers.py:
```diff
@@ -111,15 +120,24 @@ def from_xml(
-                    cls.from_xml(
-                        nested_container, ns=ns, tree=tree, parameter_lookup=parameter_lookup
+
+                if entry.attrib['containerRef'] in container_lookup:
+                    nested_container = container_lookup[entry.attrib['containerRef']]
+                else:
+                    nested_container_element = cls._find_container_by_name(
+                        tree,
+                        name=entry.attrib['containerRef'],
+                        ns=ns
                     )
+                    nested_container = cls.from_xml(
+                            nested_container_element,
+                            tree=tree,
+                            parameter_lookup=parameter_lookup,
+                            container_lookup=container_lookup,
+                            ns=ns
+                        )
```
Is there a way to change this logic at all and put the if/else into the `_find_container_by_name`? Make that a `_get_container(name, ...)` lookup that would check the cache first and it would also create the object if it isn't found. I'm not sure if that is feasible given it is a classmethod currently and doesn't have the `self.containers` information to lookup in.
There is one other place this is used, and I wans't sure it would be helpful or not to change over to `_get_container_by_name()` so just putting this out there for your thoughts.

medley56 replied:
I'm kind of trying to keep the finding of an Element separate from the parsing of that element. I've renamed these methods to `_get_base_container_element` and `_get_container_element` to make it clear they are just finding XML Element objects in the tree. `from_xml` is responsible for deciding if we actually want to parse those elements and store them in `container_lookup`.

There are probably more optimizations to make here but I want to move on for now.

greglucas commented on space_packet_parser/xtce/definitions.py:
```diff
@@ -74,16 +74,18 @@ def __init__(
-        self._sequence_container_cache = {}
+        self.parameter_types = {}
+        self.parameters = {}
+        self.containers = {}
```
Is `containers` the right name for this, or do we want to follow the XTCE naming and keep it as `sequence_containers`? I'm indifferent, these are honestly just things used internally and will rarely be used by an end-user.

medley56 replied:
I chose `containers` over `sequence_containers` because it's more concise and it contains the objects parsed out of the `<ContainerSet>` element. I just checked the element description and `SequenceContainer` is the only valid child of `ContainerSet` so I think we could use either semantic.

greglucas commented on space_packet_parser/xtce/definitions.py:
```diff
@@ -74,16 +74,18 @@ def __init__(
-                self._parameter_type_cache.update({p.parameter_type.name: p.parameter_type for p in sc.entry_list})
-                self._parameter_cache.update({p.name: p for p in sc.entry_list})
-            self._sequence_container_cache.update({sc.name: sc for sc in sequence_container_list})
+                # TODO: entry_list can contain SequenceContainers too (parseables) so this needs an
+                #  isinstance check.
+                self.parameter_types.update({p.parameter_type.name: p.parameter_type for p in sc.entry_list})
+                self.parameters.update({p.name: p for p in sc.entry_list})
+            self.containers.update({sc.name: sc for sc in sequence_container_list})
```
You are looping multiple times over the same lists, so maybe just do things explicitly with nested loops for simplicity?
```suggestion
                # entry_list can contain other containers, so filter those out and only add parameters
                for p in sc.entry_list if isinstance(p, parameters.Parameter):
                    if p.name not in self.parameters:
                        self.parameters[p.name] = p
                    if p.parameter_type.name not in self.parameter_types:
                        self.parameter_types[p.parameter_type.name] = p.parameter_type
                if sc.name not in self.containers:
                    self.containers[sc.name] = sc
```

medley56 replied:
I actually already reworked this basically as you suggest, in order to raise exceptions for `Parameter`, `ParameterType`, and `SequenceContainer` objects of the same name that are equal (`a == b`) but are not the same object in memory (`a is b`).

greglucas commented on space_packet_parser/xtce/parameter_types.py:
```diff
@@ -238,7 +241,8 @@ def from_xml(
             ns: dict,
             tree: Optional[ElementTree.Element] = None,
             parameter_lookup: Optional[dict[str, any]] = None,
-            parameter_type_lookup: Optional[dict[str, any]] = None
+            parameter_type_lookup: Optional[dict[str, any]] = None,
+            container_lookup: Optional[dict[str, any]] = None
```
These seem repetitive if they are rarely used. Would it make sense to pass in a `definition` here instead so you can consolidate all 3 of those items onto the parent object that holds all of those. That may make testing harder though?

medley56 replied:
Yeah, these are all on the ABC to satisfy Liskov substitution. The argument for keeping these kwargs separate is to make it explicit which methods use which of the three lookups. And yes, it does make things easier to test if we can just pass lists rather than a full XtcePacketDefinition object.

medley56 commented on tests/integration/test_xtce_based_parsing/test_inheritance_restrictions.py:
```diff
@@ -9,6 +9,15 @@ def test_jpss_xtce_packet_parsing(jpss_test_data_dir):
     jpss_definition = definitions.XtcePacketDefinition.from_xtce(xtce_document=jpss_xtce)
     assert isinstance(jpss_definition, definitions.XtcePacketDefinition)
 
+    # Prove that parsed objects are referencing the same objects, not duplicates
+    unused_secondary_header_container_ref = jpss_definition.containers["UNUSED"].entry_list[0]
+    assert isinstance(unused_secondary_header_container_ref, containers.SequenceContainer)
+    jpss_att_ephem_header_container_ref = jpss_definition.containers["JPSS_ATT_EPHEM"].entry_list[0]
+    assert isinstance(jpss_att_ephem_header_container_ref, containers.SequenceContainer)
+    assert unused_secondary_header_container_ref in jpss_definition.containers.values()
+    assert jpss_att_ephem_header_container_ref in jpss_definition.containers.values()
+    assert unused_secondary_header_container_ref is jpss_att_ephem_header_container_ref
```
Agreed. This is really a test for a specific regression.

medley56 commented:
@greglucas I profiled creation of the CTIM XTCE definition from XML and it actually got even faster, but that might be the hardware.

```
It took 0.21 seconds to parse the XTCE document
Number of containers: 39
Number of parameters: 9493
Number of parameter types: 15

Process finished with exit code 0
```

greglucas commented:
Yeah, that is fine with me. Just making sure we didn't go back to the 10+ seconds case. Thanks for profiling!

medley56 commented:
@greglucas All of that added time was due to the checking for unused parameters and parameter types. I removed that checking since it really isn't an error, just an indication that you aren't being very careful with your XTCE. I'm thinking we can instead add that kind of checking to whatever validator we land on out of #129 

I also added a benchmark test for this. We're back down to .03s

greglucas commented on space_packet_parser/common.py:
```diff
@@ -42,6 +42,7 @@ def from_xml(
             tree: Optional[ElementTree.ElementTree],
             parameter_lookup: Optional[dict[str, any]],
             parameter_type_lookup: Optional[dict[str, any]],
+            container_lookup: Optional[dict[str, any]],
```
Does `any` actually work with mypy, I thought it wanted `typing.Any`.

A few minor nits/questions:
* Do we want to call these "parameters", "parameter_types", "containers" so it is equivalent to what is stored on the definition? (Thinking of this now before we make a major version release)
* Should these be typed as `dict[str, Parameter]`, `dict[str, ParameterType]`, `dict[str, Parameter | SequenceContainer]`
* Since many of these are repeated all over the place, do you want to bring any of these types into a typing helpers / commons import area so you don't have to redefine them all the time?

medley56 replied:
I believe `any` is valid the same way `dict` and `list` are valid. There's also a `typing.Any` similar to `typing.Dict` and `typing.List`.

Let's address this in a future PR.

medley56 replied:
...Before we release a new major version.

greglucas replied:
`any` is a function, not a type though. I've had mypy complain in the past for any (maybe we need a mypy type-checking action too), but perhaps they've changed the typing landscape now.

greglucas reviewed (approved):
Looks good to me. I have one minor nit comment that can be addressed in a follow-up PR if we so desire to do any of those things.
